### PR TITLE
Catshadow.garbage collect conversations

### PR DESCRIPTION
**Pull request by threebithacker**
To be rebased on `main` after merging https://github.com/katzenpost/katzenpost/pull/223
- This makes sure we don't leave `LastMessage = nil` when there are live messages still in the conversation
- and memoizes the `time.Now()` arithmetic to slightly speed up things

### CODE REVIEW

**Review by mixmasala — COMMENTED** [review]
Are we sure that the changes in Start() don't leave LastMessage

File: catshadow/client.go [hunk 1]
```diff
@@ -321,12 +316,25 @@ func (c *Client) garbageCollectConversations() {
...
+					// instead of nil, should we set it to
+					// the last live message?
+					contact.LastMessage = lastLive
 				}
 				delete(messages, mesgID)
+			} else {
+				// since we aren't iterating in sorted order, we
+				// need to compare before assignment:
+				if lastLive == nil || lastLive.Timestamp.Before(message.Timestamp) {
+					lastLive = message
```
[the remark this answers is not in the document]
Comment (threebithacker) [reply catshadow/client.go:336]: @mixmasala yes indeed, you are correct!

[the remark this answers is not in the document]
Comment (threebithacker) [reply catshadow/client.go:345]: now it is :)

**Comment by threebithacker** [comment]
> I'm not sure that serializing LastMessage to/from State and removing it from Start() will work.

I think now it should be correct after `garbageCollectConversations` either way.

**Review by mixmasala — APPROVED** [review]

**Pull request merged**